Dolly 15k instruction finetuning dataset

rasbt opened the pull request:
Implementing the Dolly dataset, which is urgently needed for the NeurIPS competition (none of our current datasets is supported atm).

Seems to work fine and looks good to me, but I need to run an additional finetuning test (GPUs are currently busy due to QLoRA experiments)

Fixes #394

carmocca commented on scripts/prepare_dolly.py:
```diff
@@ -0,0 +1,170 @@
+    print("Loading data file...")
+    download_if_missing(data_file_path, data_file_url)
+    
+    with open(data_file_path, "r") as file:
```
Windows users will need 
```suggestion
    with open(data_file_path, "r", encosing="utf-8") as file:
```

carmocca commented on scripts/prepare_dolly.py:
```diff
@@ -0,0 +1,168 @@
+DATA_FILE_NAME = "dolly_data_cleaned.json"
+DESTINATION_PATH = Path("data/dolly")
+CHECKPOINT_DIR = Path("checkpoints/stabilityai/stablelm-base-alpha-3b")
+TEST_SPLIT_FRACTION = 0.1333  # to get exactly 2000 test samples
```
We were setting exactly 2k test samples for compatibility with previous experiments, but in this case we could do 10% directly

```suggestion
TEST_SPLIT_FRACTION = 0.1
```

rasbt commented:
Thanks @carmocca . I incorporated your two suggestions. 

- changing the test size to 10%
- specifying utf-8 encoding

In addition, I added an optional `--max_seq_length` argument which is useful when training on single GPU regarding memory limitations. Otherwise, most people won't be able to use it.  By default it uses the max supported sequence length from the model config like alpaca.
